### ISSUE TITLE
Thread safe, use in web site

### ISSUE THREAD

**Issue by AlexeyBoiko** [litedb-org/LiteDB#1412]
Hello,
Could you please show example of using liteDB in web site with single site instance.

As I understand it is not possible to use liteDB single file from two different processes (for example two consoles).

Is it possible to use liteDB in one process (web site) with many threads, have I to somehow register it as single instance?

**Comment by JensSchadron** [comment]
Could you give us some more information, like which version you are using (or want to use), is it in the front-end (blazer maybe?) or in the back-end and other information that might give us a bit more context of your situation.

**Comment by AlexeyBoiko** [comment]
I plan to use last version. In. .net core mvc site on a backend. I afraid many simultaneous requests can lead to locks.

**Comment by JKamsker** [comment]
Yes, a single process with many threads is exactly the intended setup. Register **one** `LiteDatabase` per database file as a singleton and share it across all requests. `LiteDatabase` is thread-safe:

- Reads don't block each other; each read sees a consistent snapshot.
- Writes are serialized per collection, so simultaneous writers wait briefly for each other instead of failing.

You can also fetch an `ILiteCollection<T>` once and keep it.

```csharp
// Program.cs
builder.Services.AddSingleton<ILiteDatabase>(_ =>
    new LiteDatabase("Filename=app.db;Connection=Direct"));
builder.Services.AddSingleton<OrdersService>();

// OrdersService.cs
public class OrdersService
{
    private readonly ILiteCollection<Order> _orders;

    public OrdersService(ILiteDatabase db) => _orders = db.GetCollection<Order>("orders");

    public List<Order> ForCustomer(string customer) =>
        _orders.Find(x => x.Customer == customer).ToList();
}
```

A few rules that avoid most problems:

- Don't create a `LiteDatabase` per request (scoped or transient). In Direct mode only one instance can have the file open, so concurrent requests would collide on it.
- Use `Connection=Shared` only if several *processes* must open the same file. It's much slower.
- Materialize query results (`ToList()`) before any `await`. Transactions belong to the thread that started them.
- Configure `BsonMapper` (`Entity<T>()...`) once at startup, not per request (#2324).
- The container disposes the singleton on shutdown, which also flushes the log file into the data file.

One caveat for 5.0.21: it has known `BsonMapper` races when several threads map a type for the first time at the same moment (#2528). The fixes are on `dev` and in the 6.x prereleases, but not in a 5.0.x NuGet release yet.

Closing this as answered — if it doesn't cover your case, comment here and we'll reopen it.
